test/test_cuda.py

```diff
@@ -5968,8 +5968,189 @@ def test_cuda_module_loading_env(self):
         self.assertEqual(val, "LAZY")
 
 
+class TestCompileKernel(TestCase):
+    @unittest.skipIf(not TEST_CUDA, "No CUDA")
+    def test_compile_kernel(self):
+        # Simple vector addition kernel
+        kernel_source = """
+        __global__ void add_tensors(const float* a, const float* b, float* c, int n) {
+            int i = threadIdx.x + blockIdx.x * blockDim.x;
+            if (i < n)
+                c[i] = a[i] + b[i];
+        }
+        """
+
+        # Compile the kernel
+        from torch.cuda import _compile_kernel
+
+        add_kernel = _compile_kernel(kernel_source, "add_tensors")
```

Comment:
From composability side: if you make it mandatory for users to specify which float* they are actually writing to, then we can automatically support these kernels with torch.compile without using the torch.library API (among other goodies, like being able to interface with PyTorch's tensor version counters). That's what we do for the Python custom operator API. cc @albanD 

Alternatively we could parse the code for if it's using const pointers or regular pointers, but I don't think PyTorch users are that disciplined.

I get that this add some more barrier for someone trying to use a CUDA kernel with pytorch so feel free to push back

Reply:
@zou3519 Indeed my primary goal has been minimal code changes to CUDA with minimal compilation times. Granted it's an interesting idea considering Triton also asks users to be explicit about output tensors

I do personally think this PR is useful enough as is to be merged and considering it's a private API I can revisit immediately next what's the right way to make this work with `torch.compile` - at the very least we could explore adding support for storing kwarg on the function so its easier for us to play around with the user annotation API.

Reply:
Yup happy to get this merged first so we can keep prototyping things on top of it